src/main/java/com/mraof/minestuck/blockentity/machine/IntellibeamLaserstationBlockEntity.java

```diff
@@ -0,0 +1,238 @@
+	{
+		if(intellibeam.waitTimer > 0)
+		{
+			intellibeam.waitTimer--;
+		}
+	}
+	
+	public void onRightClick(Player player)
+	{
+		if(level != null && !level.isClientSide)
+		{
+			ItemStack heldCard = player.getMainHandItem();
+			ItemStack itemInHeldCard = AlchemyHelper.getDecodedItem(heldCard);
+			
+			if(waitTimer > 0)
+			{
+				return;
+			}
+			
+			if(!analyzedCard.isEmpty() && player.isShiftKeyDown())
+			{
+				takeCard(player);
+				waitTimer = 10;
+			} else if(AlchemyHelper.isReadableCard(analyzedCard))
+			{
+				this.level.playSound(null, this.worldPosition, MSSoundEvents.INTELLIBEAM_LASERSTATION_REMOVE_CARD.get(), SoundSource.BLOCKS, 0.5F, 0.1F);
+				player.displayClientMessage(Component.translatable(CAPTCHA_DECODED), true);
+				waitTimer = 10;
+			} else if(analyzedCard.isEmpty() && itemInHeldCard.is(MSTags.Items.UNREADABLE))
+			{
+				setCard(heldCard.split(1));
+				waitTimer = 10;
+			} else if(getCardItemExperience() >= EXP_LEVEL_CAPACITY)
+			{
+				applyReadableNBT(analyzedCard);
+				takeCard(player);
+				
+				waitTimer = 10;
+			} else
+			{
+				addExperience(player);
+				waitTimer = 10;
+			}
+		}
+	}
+	
+	private Integer getCardItemExperience()
+	{
+		return decodingProgress.getOrDefault(AlchemyHelper.getDecodedItem(analyzedCard).getItem(), 0);
+	}
+	
+	public String processExperienceGuage()
+	{
+		int numberOfBars = "▯▯▯▯▯▯▯▯▯▯".length(); //used to visualize how many experience bars there are
+		
+		char[] stringChars = new char[numberOfBars];
+		
+		for(int i = 0; i < numberOfBars; i++)
+		{
+			if(i < getCardItemExperience())
+				stringChars[i] = '▮';
+			else
+				stringChars[i] = '▯';
+		}
+		
+		return '<' + new String(stringChars) + '>'; //if two levels of experience have been invested, it will look like this: "<▮▮▯▯▯▯▯▯▯▯>"
+	}
+	
+	public void takeCard(Player player)
+	{
+		if(player.getMainHandItem().isEmpty())
+			player.setItemInHand(InteractionHand.MAIN_HAND, analyzedCard);
+		else if(!player.getInventory().add(analyzedCard))
+			dropCard(false, level, worldPosition, analyzedCard);
+		else player.inventoryMenu.broadcastChanges();
+		
+		this.level.playSound(null, this.worldPosition, MSSoundEvents.INTELLIBEAM_LASERSTATION_REMOVE_CARD.get(), SoundSource.BLOCKS, 0.5F, 1F);
+		
+		setCard(ItemStack.EMPTY);
+	}
+	
+	public void dropCard(boolean inBlock, Level level, BlockPos pos, ItemStack item)
+	{
+		BlockPos dropPos;
+		if(inBlock)
+			dropPos = pos;
+		else if(!Block.canSupportCenter(level, pos.above(), Direction.DOWN))
+			dropPos = pos.above();
+		else dropPos = pos;
+		
+		Containers.dropItemStack(level, dropPos.getX(), dropPos.getY(), dropPos.getZ(), item);
+	}
+	
+	public void setCard(ItemStack card)
+	{
+		if(card.is(MSItems.CAPTCHA_CARD.get()) || card.isEmpty())
+		{
+			this.analyzedCard = card;
+			if(level != null)
+			{
+				updateState();
+			}
+		}
+	}
+	
+	public void applyReadableNBT(ItemStack taggedCard)
+	{
+		taggedCard.getOrCreateTag().putBoolean("decoded", true);
```

Comment:
I suppose we don't need "decoded" anymore, and could instead check whether the stack has the "captcha_code" tag or now?